File: src/main/java/kr/mjc/rachel/concepts/functions/ConsumerExample.java
/*
 * @FunctionalInterface : abstract 메서드가 1개 있는 인터페이스
 * Consumer<T> : 인자가 1개 있고, 리턴이 없는 함수를 갖는 @FunctionalInterface
 */
package kr.mjc.rachel.concepts.functions;

import java.util.List;
import java.util.function.Consumer;

public class ConsumerExample {
  /**
   * list의 요소에 consumer.accept(t)를 실행한다.
   */
  public <T> void repeat(List<T> list, Consumer<T> consumer) {
    for (T t : list)
      consumer.accept(t);
  }

  public static void main(String[] args) {
    ConsumerExample exam = new ConsumerExample();
    List<String> list1 = List.of("apple", "tomato", "lemon");

    Consumer<String> c1 = new Consumer<>() { // anonymous class
      @Override
      public void accept(String str) {
        System.out.println(str);
      }
    };

    exam.repeat(list1, c1);
  }
}

File: src/main/kotlin/kr/mjc/jacob/concepts/functions/ConsumerExample.kt
package kr.mjc.jacob.concepts.functions

class ConsumerExample {
  // higher order function
  // inline 함수 인스턴스를 만들지 않고 함수의 코드를 삽입한다.
  inline fun <T> repeat(list: List<T>, accept: (T) -> Unit) {
    for (e in list) accept(e)
  }

  fun printString(str: String) {
    println("$str, length=${str.length}")
  }
}

fun main() {
  val exam = ConsumerExample()
  val list1 = listOf("apple", "tomato", "lemon")

  // lamda expression
  // val 변수명 : (인자타입) -> 리턴타입 = { 인자 -> 구현 }
  val c1: (String) -> Unit = { e -> println(e) }

  exam.repeat(list1, c1)
  exam.repeat(list1) { e -> println(e) }
  exam.repeat(list1, exam::printString)
}
